R8 fullMode removes user flag interfaces / FlagSource subclasses — Flagfit.create<T>() fails with ClassCastException

## Environment

- flagfit: 1.1.7 (and flagfit-flagtype: 1.1.7)
- AGP: 8.13.2
- R8: shipped with AGP 8.13.2
- `android.enableR8.fullMode=true` (AGP 8.0+ default)
- Kotlin: 2.2.21
- Affected app: an internal Android app (not publicly linkable); a minimal repro repository can be provided on request.

## Summary

With R8 fullMode enabled, `Flagfit.create<MyFlagService>()` throws `ClassCastException` at application start because the interface methods / FlagSource subclasses are removed by R8's tree-shaking. The consumer ProGuard rules currently shipped in `flagfit/src/main/resources/META-INF/proguard/flagfit.pro` only keep annotation classes, not the user-side interface or the `FlagSource` subclasses referenced by `@DebugWith` / `@ReleaseWith` / `@DefaultWith`.

## Reproduction

1. Create an Android app with AGP 8.13 and Kotlin 2.2.
2. Add `flagfit:1.1.7` and `flagfit-flagtype:1.1.7`.
3. Define a flag service interface and call `Flagfit().create<MyFlagService>()` during `Application.onCreate()`.
4. Enable `isMinifyEnabled = true` and `android.enableR8.fullMode=true` (default in AGP 8+).
5. `./gradlew :app:bundleRelease` and launch the resulting APK.

Expected: app launches normally.
Actual: `java.lang.ClassCastException` inside the chained injection path:

```
Caused by: java.lang.ClassCastException
  at com.example.MyRepository.<init>(MyRepository.kt:NN)
  at com.example.App.onCreate(App.kt:NN)
```

After retrace, the failing site is `flagfit.create<MyFlagService>()` — the cast of the JDK proxy to the interface type fails because the interface's methods have been shrunk by fullMode.

## Root cause

`Flagfit.create<T>()` uses `Proxy.newProxyInstance` with reflection over the interface's `declaredMethods` and their annotations. Under R8 fullMode:

- The user interface (e.g. `MyFlagService`) is not kept; methods are stripped.
- Annotations defined in `flagfit-flagtype` (`FlagType.Experiment`, etc.) are not covered by any consumer rules — that submodule has no `META-INF/proguard/*.pro` at all.
- `FlagSource` subclasses referenced only via `@DebugWith(X::class)` are removed because R8 does not see a direct reference.

The current consumer rules (`flagfit.pro`) only keep annotation classes themselves; that is sufficient for the legacy non-fullMode behaviour but not for fullMode.

## Proposed fix (upstream)

### 1. Extend `flagfit/src/main/resources/META-INF/proguard/flagfit.pro`

```proguard
# Keep the annotation types (existing)
-keep class tv.abema.flagfit.annotation.**
-keep class tv.abema.flagfit.SuspendReturnType

# Annotations must survive with their runtime values so that
# Flagfit.create() can read them via reflection.
-keepattributes RuntimeVisibleAnnotations,AnnotationDefault,Signature

# Any interface that has at least one flagfit-annotated method must be kept
# entirely; Flagfit.create() produces a JDK Proxy implementing that interface
# and calls declaredMethods on it.
-keep,allowobfuscation interface * {
    @tv.abema.flagfit.annotation.BooleanFlag <methods>;
}
-keep,allowobfuscation interface * {
    @tv.abema.flagfit.annotation.VariationFlag <methods>;
}
-keepclassmembers,allowobfuscation interface * {
    @tv.abema.flagfit.annotation.BooleanFlag <methods>;
    @tv.abema.flagfit.annotation.VariationFlag <methods>;
    @tv.abema.flagfit.annotation.BooleanEnv <methods>;
    @tv.abema.flagfit.annotation.DebugWith <methods>;
    @tv.abema.flagfit.annotation.ReleaseWith <methods>;
    @tv.abema.flagfit.annotation.DefaultWith <methods>;
}

# FlagSource subclasses referenced only from @DebugWith / @ReleaseWith /
# @DefaultWith cannot be tree-shaken.
-keep class * extends tv.abema.flagfit.FlagSource
```

### 2. Add `flagfit-flagtype/src/main/resources/META-INF/proguard/flagfit-flagtype.pro` (new)

```proguard
# FlagType sub-annotations used to classify flags
-keep class tv.abema.flagfit.FlagType
-keep class tv.abema.flagfit.FlagType$* { *; }
```

### 3. README.md — add an "R8 / ProGuard" section

Note that the bundled consumer rules now cover fullMode; no extra app-side configuration is required. Explicitly mention support for AGP 8+ / fullMode.

## Alternative (for library consumers until upstream ships)

Add to the consuming app's `proguard-rules.pro`:

```proguard
-keepattributes RuntimeVisibleAnnotations,AnnotationDefault,Signature
-keep interface <your flag service package>.** { *; }
-keep class * extends tv.abema.flagfit.FlagSource
-keep class tv.abema.flagfit.FlagType$* { *; }
```

## Why fix upstream rather than only in consumers

1. The current bundled `flagfit.pro` already signals intent to support ProGuard.
2. fullMode is the AGP 8+ default; silently breaking consumers is a regression.
3. `flagfit-flagtype` has no rules at all — consumer-side workarounds are easy to get wrong (I initially only added rules for `annotation.**` and missed `FlagType.*`).
4. Retrofit ships the equivalent fullMode rules (`-if interface * { @retrofit2.http.* <methods>; } -keep,allowobfuscation interface <1>`) in its own consumer rules since 2.9.0; flagfit follows the same proxy pattern and should do the same.

## References

- flagfit `Flagfit.kt` (source of the Proxy.newProxyInstance call): https://github.com/abema/flagfit/blob/main/flagfit/src/main/java/tv/abema/flagfit/Flagfit.kt
- Current consumer rules: https://github.com/abema/flagfit/blob/main/flagfit/src/main/resources/META-INF/proguard/flagfit.pro
- Retrofit's equivalent rules (precedent for this pattern): https://github.com/square/retrofit/blob/master/retrofit/src/main/resources/META-INF/proguard/retrofit2.pro

--- a/README.md
+++ b/README.md
@@ -487,12 +487,18 @@ If you can't yet upgrade to a version that ships these rules, add the following 
 your app's `proguard-rules.pro`:
 
 ```proguard
--keepattributes RuntimeVisibleAnnotations,AnnotationDefault,Signature
--keep class tv.abema.flagfit.annotation.**
--keep class tv.abema.flagfit.SuspendReturnType
--keep class tv.abema.flagfit.FlagType
--keep class tv.abema.flagfit.FlagType$* { *; }
--keep class * extends tv.abema.flagfit.FlagSource
+-keepattributes RuntimeVisibleAnnotations
+
+-keep,allowobfuscation class tv.abema.flagfit.annotation.BooleanFlag
+-keep,allowobfuscation class tv.abema.flagfit.annotation.VariationFlag
+-keep,allowobfuscation class tv.abema.flagfit.annotation.BooleanEnv
+-keep,allowobfuscation class tv.abema.flagfit.annotation.DebugWith
+-keep,allowobfuscation class tv.abema.flagfit.annotation.ReleaseWith
+-keep,allowobfuscation class tv.abema.flagfit.annotation.DefaultWith
+-keep,allowobfuscation class tv.abema.flagfit.SuspendReturnType
+-keep,allowobfuscation class tv.abema.flagfit.FlagType
+-keep,allowobfuscation class tv.abema.flagfit.FlagType$* { *; }
+-keep,allowobfuscation class * extends tv.abema.flagfit.FlagSource
 
 -if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
 -keep,allowobfuscation interface <1>

--- a/README.md
+++ b/README.md
@@ -462,6 +462,44 @@ val flagfit = Flagfit(
 )
 ```
 
+## R8 / ProGuard
+
+Flagfit uses `Proxy.newProxyInstance` and reads each interface method's annotations
+via reflection at runtime. To keep this working under code shrinking — including
+**R8 fullMode**, which is the default since AGP 8.0 — both `flagfit` and
+`flagfit-flagtype` ship consumer ProGuard rules in their `META-INF/proguard/`
+directory.
+
+If you depend on `flagfit:1.1.8` (or newer) and `flagfit-flagtype:1.1.8` (or newer)
+no extra app-side configuration is required — the bundled rules already keep:
+
+- the flagfit annotation classes (`@BooleanFlag`, `@VariationFlag`, `@BooleanEnv`,
+  `@DebugWith`, `@ReleaseWith`, `@DefaultWith`) and their runtime values,
+- any interface that declares a flagfit-annotated method (so `Flagfit.create()`'s
+  proxy + `declaredMethods` reflection still works after shrinking),
+- every `FlagSource` subclass referenced via `@DebugWith` / `@ReleaseWith` /
+  `@DefaultWith` annotation values,
+- `FlagType` and all its inner annotation / `AnnotationAdapter` classes.
+
+### Workaround for older versions
+
+If you can't yet upgrade to a version that ships these rules, add the following to
+your app's `proguard-rules.pro`:
+
+```proguard
+-keepattributes RuntimeVisibleAnnotations,AnnotationDefault,Signature
+-keep class tv.abema.flagfit.annotation.**
+-keep class tv.abema.flagfit.SuspendReturnType
+-keep class tv.abema.flagfit.FlagType
+-keep class tv.abema.flagfit.FlagType$* { *; }
+-keep class * extends tv.abema.flagfit.FlagSource
+
+-if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
+-keep,allowobfuscation interface <1>
+-if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
+-keep,allowobfuscation interface <1>
+```
+
 ## Lint check based on expiration date
 
 Flags that have passed their expiration date or are scheduled to expire within the next 7 days will be displayed as warnings in the IDE.

--- a/README.md
+++ b/README.md
@@ -464,64 +464,9 @@ val flagfit = Flagfit(
 
 ## R8 / ProGuard
 
-Flagfit uses `Proxy.newProxyInstance` and reads each interface method's annotations
-via reflection at runtime. To keep this working under code shrinking — including
-**R8 fullMode**, which is the default since AGP 8.0 — both `flagfit` and
-`flagfit-flagtype` ship consumer ProGuard rules in their `META-INF/proguard/`
-directory.
+If you are using R8 the shrinking and obfuscation rules are included automatically.
 
-If you depend on `flagfit:1.1.8` (or newer) and `flagfit-flagtype:1.1.8` (or newer)
-no extra app-side configuration is required — the bundled rules already keep:
-
-- the flagfit annotation classes (`@BooleanFlag`, `@VariationFlag`, `@BooleanEnv`,
-  `@DebugWith`, `@ReleaseWith`, `@DefaultWith`) and their runtime values,
-- any interface that declares a flagfit-annotated method (so `Flagfit.create()`'s
-  proxy + `declaredMethods` reflection still works after shrinking),
-- every `FlagSource` subclass referenced via `@DebugWith` / `@ReleaseWith` /
-  `@DefaultWith` annotation values,
-- `FlagType` and all its inner annotation / `AnnotationAdapter` classes.
-
-### Workaround for older versions
-
-If you can't yet upgrade to a version that ships these rules, add the following to
-your app's `proguard-rules.pro`:
-
-```proguard
--keepattributes RuntimeVisibleAnnotations
-
--keep,allowobfuscation class tv.abema.flagfit.annotation.BooleanFlag
--keep,allowobfuscation class tv.abema.flagfit.annotation.VariationFlag
--keep,allowobfuscation class tv.abema.flagfit.annotation.BooleanEnv
--keep,allowobfuscation class tv.abema.flagfit.annotation.DebugWith
--keep,allowobfuscation class tv.abema.flagfit.annotation.ReleaseWith
--keep,allowobfuscation class tv.abema.flagfit.annotation.DefaultWith
--keep,allowobfuscation class tv.abema.flagfit.SuspendReturnType
--keep,allowobfuscation class tv.abema.flagfit.FlagType
--keep,allowobfuscation class tv.abema.flagfit.FlagType$* { *; }
--keep,allowobfuscation class * extends tv.abema.flagfit.FlagSource
-
--if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
--keep,allowobfuscation interface <1>
--if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
--keep,allowobfuscation interface <1>
-
-# Subinterfaces of an annotated flag service must also survive R8 fullMode.
--if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
--keep,allowobfuscation interface * extends <1>
--if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
--keep,allowobfuscation interface * extends <1>
-
-# Flagfit reads each annotated method's annotations via reflection, so the
-# methods themselves must be preserved on every flag-service interface.
--keepclassmembers,allowobfuscation interface * {
-    @tv.abema.flagfit.annotation.BooleanFlag <methods>;
-    @tv.abema.flagfit.annotation.VariationFlag <methods>;
-    @tv.abema.flagfit.annotation.BooleanEnv <methods>;
-    @tv.abema.flagfit.annotation.DebugWith <methods>;
-    @tv.abema.flagfit.annotation.ReleaseWith <methods>;
-    @tv.abema.flagfit.annotation.DefaultWith <methods>;
-}
-```
+ProGuard users must manually add the options from [flagfit.pro](flagfit/src/main/resources/META-INF/proguard/flagfit.pro) and [flagfit-flagtype.pro](flagfit-flagtype/src/main/resources/META-INF/proguard/flagfit-flagtype.pro).
 
 ## Lint check based on expiration date
 

--- a/README.md
+++ b/README.md
@@ -498,6 +498,23 @@ your app's `proguard-rules.pro`:
 -keep,allowobfuscation interface <1>
 -if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
 -keep,allowobfuscation interface <1>
+
+# Subinterfaces of an annotated flag service must also survive R8 fullMode.
+-if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
+-keep,allowobfuscation interface * extends <1>
+-if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
+-keep,allowobfuscation interface * extends <1>
+
+# Flagfit reads each annotated method's annotations via reflection, so the
+# methods themselves must be preserved on every flag-service interface.
+-keepclassmembers,allowobfuscation interface * {
+    @tv.abema.flagfit.annotation.BooleanFlag <methods>;
+    @tv.abema.flagfit.annotation.VariationFlag <methods>;
+    @tv.abema.flagfit.annotation.BooleanEnv <methods>;
+    @tv.abema.flagfit.annotation.DebugWith <methods>;
+    @tv.abema.flagfit.annotation.ReleaseWith <methods>;
+    @tv.abema.flagfit.annotation.DefaultWith <methods>;
+}
 ```
 
 ## Lint check based on expiration date

--- a/flagfit-flagtype/src/main/resources/META-INF/proguard/flagfit-flagtype.pro
+++ b/flagfit-flagtype/src/main/resources/META-INF/proguard/flagfit-flagtype.pro
@@ -1,6 +1,8 @@
-# FlagType outer class and all inner classes (FlagType.WorkInProgress, .Experiment,
-# .Ops, .Permission and the four *AnnotationAdapter inner classes) must survive
-# because Flagfit.create() reflects on the annotations and FlagType.annotationAdapters()
-# instantiates the adapters by name.
--keep class tv.abema.flagfit.FlagType
--keep class tv.abema.flagfit.FlagType$* { *; }
+# FlagType outer class and all inner classes (FlagType.WorkInProgress,
+# .Experiment, .Ops, .Permission and the four *AnnotationAdapter inner
+# classes) must survive because Flagfit.create() reflects on the annotations
+# and FlagType.annotationAdapters() instantiates the adapters.
+# Class literal references in user code are rewritten by R8 in lock-step,
+# so allowobfuscation is safe.
+-keep,allowobfuscation class tv.abema.flagfit.FlagType
+-keep,allowobfuscation class tv.abema.flagfit.FlagType$* { *; }

--- a/flagfit-flagtype/src/main/resources/META-INF/proguard/flagfit-flagtype.pro
+++ b/flagfit-flagtype/src/main/resources/META-INF/proguard/flagfit-flagtype.pro
@@ -1,0 +1,6 @@
+# FlagType outer class and all inner classes (FlagType.WorkInProgress, .Experiment,
+# .Ops, .Permission and the four *AnnotationAdapter inner classes) must survive
+# because Flagfit.create() reflects on the annotations and FlagType.annotationAdapters()
+# instantiates the adapters by name.
+-keep class tv.abema.flagfit.FlagType
+-keep class tv.abema.flagfit.FlagType$* { *; }

--- a/flagfit/src/main/resources/META-INF/proguard/flagfit.pro
+++ b/flagfit/src/main/resources/META-INF/proguard/flagfit.pro
@@ -16,6 +16,15 @@
 -if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
 -keep,allowobfuscation interface <1>
 
+# Also keep subinterfaces of the above, because callers may proxy a child
+# interface (e.g. `interface Child : Parent`) whose annotated methods live on
+# the parent. R8 fullMode would otherwise be free to strip / merge `Child`.
+-if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
+-keep,allowobfuscation interface * extends <1>
+
+-if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
+-keep,allowobfuscation interface * extends <1>
+
 # Keep the annotated methods (and their annotations) on those interfaces so
 # Method.getAnnotations() returns the real values.
 -keepclassmembers,allowobfuscation interface * {

--- a/flagfit/src/main/resources/META-INF/proguard/flagfit.pro
+++ b/flagfit/src/main/resources/META-INF/proguard/flagfit.pro
@@ -1,2 +1,32 @@
+# Keep flagfit annotation classes themselves.
 -keep class tv.abema.flagfit.annotation.**
 -keep class tv.abema.flagfit.SuspendReturnType
+
+# Annotations must survive with their runtime values so Flagfit.create() can read
+# them via reflection at runtime.
+-keepattributes RuntimeVisibleAnnotations,AnnotationDefault,Signature
+
+# Flagfit.create() builds a JDK Proxy for any interface whose methods are annotated
+# with @BooleanFlag or @VariationFlag, then iterates declaredMethods. R8 fullMode
+# removes such interfaces because it sees no concrete implementation, so explicitly
+# keep them. (Same pattern Retrofit uses for @retrofit2.http.* methods.)
+-if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
+-keep,allowobfuscation interface <1>
+
+-if interface * { @tv.abema.flagfit.annotation.VariationFlag <methods>; }
+-keep,allowobfuscation interface <1>
+
+# Keep the annotated methods (and their annotations) on those interfaces so
+# Method.getAnnotations() returns the real values.
+-keepclassmembers,allowobfuscation interface * {
+    @tv.abema.flagfit.annotation.BooleanFlag <methods>;
+    @tv.abema.flagfit.annotation.VariationFlag <methods>;
+    @tv.abema.flagfit.annotation.BooleanEnv <methods>;
+    @tv.abema.flagfit.annotation.DebugWith <methods>;
+    @tv.abema.flagfit.annotation.ReleaseWith <methods>;
+    @tv.abema.flagfit.annotation.DefaultWith <methods>;
+}
+
+# FlagSource subclasses referenced only via @DebugWith / @ReleaseWith / @DefaultWith
+# (KClass<out FlagSource> annotation values) are otherwise unreachable.
+-keep class * extends tv.abema.flagfit.FlagSource

--- a/flagfit/src/main/resources/META-INF/proguard/flagfit.pro
+++ b/flagfit/src/main/resources/META-INF/proguard/flagfit.pro
@@ -1,15 +1,31 @@
-# Keep flagfit annotation classes themselves.
--keep class tv.abema.flagfit.annotation.**
--keep class tv.abema.flagfit.SuspendReturnType
+# Keep flagfit annotation classes individually (avoid library-package wildcard
+# per the Android library-optimization guidance). Lookups in Flagfit.kt use
+# class literals which R8 rewrites in lock-step, so allowobfuscation is safe.
+-keep,allowobfuscation class tv.abema.flagfit.annotation.BooleanFlag
+-keep,allowobfuscation class tv.abema.flagfit.annotation.VariationFlag
+-keep,allowobfuscation class tv.abema.flagfit.annotation.BooleanEnv
+-keep,allowobfuscation class tv.abema.flagfit.annotation.DebugWith
+-keep,allowobfuscation class tv.abema.flagfit.annotation.ReleaseWith
+-keep,allowobfuscation class tv.abema.flagfit.annotation.DefaultWith
+-keep,allowobfuscation class tv.abema.flagfit.SuspendReturnType
 
-# Annotations must survive with their runtime values so Flagfit.create() can read
-# them via reflection at runtime.
--keepattributes RuntimeVisibleAnnotations,AnnotationDefault,Signature
+# Annotation values must survive at runtime so Flagfit.create() can read them
+# via reflection. Only RuntimeVisibleAnnotations belongs in consumer rules per
+# the Android library-optimization guidance — Flagfit does not read annotation
+# defaults or generic signatures, so AnnotationDefault and Signature are not
+# needed here.
+-keepattributes RuntimeVisibleAnnotations
 
-# Flagfit.create() builds a JDK Proxy for any interface whose methods are annotated
-# with @BooleanFlag or @VariationFlag, then iterates declaredMethods. R8 fullMode
-# removes such interfaces because it sees no concrete implementation, so explicitly
-# keep them. (Same pattern Retrofit uses for @retrofit2.http.* methods.)
+# Flagfit.create() builds a JDK Proxy for any interface whose methods are
+# annotated with @BooleanFlag or @VariationFlag, then iterates declaredMethods.
+# R8 fullMode removes such interfaces because it sees no concrete
+# implementation, so explicitly keep them.
+#
+# NOTE: do NOT add ,allowshrinking here. Although Retrofit ships its
+# equivalent rule with allowshrinking, empirically adding it causes R8
+# fullMode to tree-shake the annotated methods on user flag-service
+# interfaces (because the only dynamic caller is Proxy.invoke via reflection,
+# which R8 does not trace), which re-introduces the issue this rule fixes.
 -if interface * { @tv.abema.flagfit.annotation.BooleanFlag <methods>; }
 -keep,allowobfuscation interface <1>
 
@@ -36,6 +52,9 @@
     @tv.abema.flagfit.annotation.DefaultWith <methods>;
 }
 
-# FlagSource subclasses referenced only via @DebugWith / @ReleaseWith / @DefaultWith
-# (KClass<out FlagSource> annotation values) are otherwise unreachable.
--keep class * extends tv.abema.flagfit.FlagSource
+# FlagSource subclasses referenced only via @DebugWith / @ReleaseWith /
+# @DefaultWith (KClass<out FlagSource> annotation values) are otherwise
+# unreachable. allowobfuscation is safe (R8 rewrites class literals); do NOT
+# add allowshrinking — under R8 fullMode KClass annotation values are not
+# always traced as live references and dropping them would re-introduce #39.
+-keep,allowobfuscation class * extends tv.abema.flagfit.FlagSource

